pyro/contrib/forecast/forecaster.py

```diff
@@ -0,0 +1,221 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+from abc import ABCMeta, abstractmethod
+
+import torch
+import torch.nn as nn
+
+import pyro
+import pyro.distributions as dist
+import pyro.poutine as poutine
+from pyro.infer import SVI, Trace_ELBO
+from pyro.infer.autoguide import AutoNormal, init_to_sample
+from pyro.nn.module import PyroModule
+from pyro.optim import ClippedAdam
+
+from .util import PrefixConditionMessenger, PrefixReplayMessenger, reshape_batch
+
+logger = logging.getLogger(__name__)
+
+
+class _ForecastingModelMeta(type(PyroModule), ABCMeta):
+    pass
+
+
+class ForecastingModel(PyroModule, metaclass=_ForecastingModelMeta):
+    """
+    Abstract base class for forecasting models.
+
+    Derived classes must implement the :meth:`model` method.
+    """
+    def __init__(self):
+        super().__init__()
+        self._prefix_condition_data = {}
+
+    @abstractmethod
+    def model(self, zero_data, covariates):
+        """
+        Generative model definition.
+
+        Implementations must call the :meth:`predict` method exactly once.
+
+        Implementations must draw all time-dependent noise inside the
+        :meth:`time_plate` .  The prediction passed to :meth:`predict` must be
```

Comment:
Are you referring to the space between the tick and the period? I always include this because sphinx [restrictions on quoting](http://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#gotchas). I don't fully understand the rules, but after having spent hours debugging subtle sphinx breakages, I now always surround tick links with space. Do you have a clearer understanding of sphinx restrictions?

Reply:
That and the double space before `The` - my keyboard has a double keystroke problem with many keys, which turns out to be a common problem with recent macbooks. I thought you might have the same issue with yours. I didn't know about the sphinx linking issue, I have faced it before and didn't realize that I could get around by putting a space after the backtick, I'll keep that in mind for next time.

Reply:
Well I've removed the spaces and the docs built fine 🤷‍♂